### PR TITLE
DrivAerML anchor budget sweep: 4096→16384 supernodes

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

DrivAerML is a 3D dataset with complex geometry. The published AB-UPT baseline uses geometry_supernodes=16384 and surface_anchor_points=16384, while our default is 4096/8000 (much smaller). Increasing anchor budgets gives the model higher resolution to represent the car surface — likely critical for beating Transolver-3's 3.71%. We test 3 anchor budget levels while keeping training feasible.

## Current Baseline

- DrivAerML: **no baseline yet** — target to beat: surface_rel_l2_pct < 3.71% (Transolver-3)

## Instructions

Run 4 parallel trials with `CUDA_VISIBLE_DEVICES`. Use `--wandb_group radford-drivaerml-round1` for all runs. AdamW lr=5e-4, batch_size=1, cosine_t_max=30.

### Trial 0 — GPU 0: Default anchors (geometry_supernodes=4096, surface_anchor_points=8000)
```bash
CUDA_VISIBLE_DEVICES=0 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine_t_max 30 --batch-size 1 --geometry-supernodes 4096 --surface-anchor-points 8000 --wandb_group radford-drivaerml-round1 --wandb_name nezuko/drivaerml-anchor-default --agent nezuko
```

### Trial 1 — GPU 1: Medium anchors (geometry_supernodes=8192, surface_anchor_points=16000)
```bash
CUDA_VISIBLE_DEVICES=1 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine_t_max 30 --batch-size 1 --geometry-supernodes 8192 --surface-anchor-points 16000 --wandb_group radford-drivaerml-round1 --wandb_name nezuko/drivaerml-anchor-medium --agent nezuko
```

### Trial 2 — GPU 2: Large anchors (geometry_supernodes=16384, surface_anchor_points=16384)
```bash
CUDA_VISIBLE_DEVICES=2 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine_t_max 30 --batch-size 1 --geometry-supernodes 16384 --surface-anchor-points 16384 --wandb_group radford-drivaerml-round1 --wandb_name nezuko/drivaerml-anchor-large --agent nezuko
```

### Trial 3 — GPU 3: High surface resolution with point budget (supernodes=8192, surface_anchor=16000, surface_points=16000)
```bash
CUDA_VISIBLE_DEVICES=3 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine_t_max 30 --batch-size 1 --geometry-supernodes 8192 --surface-anchor-points 16000 --drivaerml-train-surface-points 16000 --drivaerml-eval-surface-points 16000 --wandb_group radford-drivaerml-round1 --wandb_name nezuko/drivaerml-anchor-med-pts16k --agent nezuko
```

Run all 4 in parallel (background each with `&`). Wait for all to finish.

**Notes:**
- If Trial 2 (geometry_supernodes=16384) OOMs, try batch_size=1 with these params — note the OOM in results
- Primary metric: `val_primary/surface_rel_l2_pct` (lower is better)
- Report GPU memory usage, epoch time, and val metric for each trial

## Results

_To be filled by student after runs complete._

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT nezuko:

## Progress update — critical findings before running trials

### Issue 1: geometry_supernodes and surface_anchor_points are no-ops for senpai_transolver

The parameters being swept (`geometry_supernodes`, `surface_anchor_points`) are only used by the `reference_abupt` model path in `build_loaders()` via `ABUPTCollate`. For `senpai_transolver`, the collate function is `collate_grouped` which ignores both parameters. The model receives all surface points from the data pipeline directly.

**Code evidence:**
- `build_loaders()` (train.py:462): `ABUPTCollate` uses `geometry_supernodes`/`surface_anchor_points` only when `config.model == "reference_abupt"`
- `build_model()` (train.py:425): `SenpaiTransolver` constructor doesn't receive either parameter

### Issue 2: DrivAerML cases have ~8.8M surface points, not ~40K

Each DrivAerML case has approximately **8,828,095 surface points** (run_1). With `drivaerml_train_surface_points=8000` (my workaround for OOM), the dataset creates `ceil(8.8M/8K) = 1,104 views per case`, yielding **421,902 training iterations per epoch**. At ~1.5s per iteration, that's **~68 hours per epoch** — completely infeasible.

The first 4 trials ran for 1+ hours without completing a single training iteration (stuck iterating through the massive DataLoader).

### Issue 3: Memory-feasible point budgets for senpai_transolver

GPU benchmarks (fwd+bwd with bs=1):
| surface pts/view | peak GPU | fwd+bwd time |
|---|---|---|
| 500K | 26.7 GB | 0.30s |
| 1M | 53.3 GB | 0.59s |
| 1.5M | 80.1 GB | 0.89s |
| 2M+ | OOM | — |

NFS load times are nearly constant (~0.55-0.73s) regardless of point count (mmap).

### Plan: Run 2 feasible trials

Since the original sweep parameters don't affect senpai_transolver, I'll run the closest meaningful analog — a **surface point budget sweep** — which controls how many points the model sees per training iteration:

- **Trial A** (GPU 0): `drivaerml_train_surface_points=500000` → 26.7GB, ~7K views/epoch
- **Trial B** (GPU 1): `drivaerml_train_surface_points=1500000` → 80GB, ~2.4K views/epoch

Both with `max-train-batches=394` to complete within timeout, `num-workers=2` for I/O overlap.

Kills were sent to all 4 original processes. Relaunching now.

**Comment by tcapelle** [comment]
STUDENT nezuko:

## Results

### Critical architectural finding

**`geometry_supernodes` and `surface_anchor_points` are no-ops for `senpai_transolver`.**

These parameters are only consumed by `ABUPTCollate` (used when `config.model == "reference_abupt"`). For `senpai_transolver`, the collate function is `collate_grouped` which passes all surface points directly to the model. The 4-trial anchor budget sweep as originally designed would have tested identical model behavior across all trials.

### DrivAerML scale discovery

Each DrivAerML case has **~8.8 million surface points** (e.g., run_1 = 8,828,095). This is ~1000x larger than initially expected. With the original `drivaerml_train_surface_points=8000`, the dataset creates **1,104 views per case** → **421,902 training iterations per epoch** → estimated **68 hours/epoch**. The first 4 trials ran 1+ hours stuck iterating through this massive DataLoader with 0% GPU utilization.

### Feasible trials: surface point budget sweep

Ran 2 trials with `max-train-batches=394` and `num-workers=2`:

| Trial | surface_pts/view | Peak GPU | Runtime | val_rel_l2_pct (ep2) | test_rel_l2_pct | W&B |
|---|---|---|---|---|---|---|
| A | 500,000 | ~27 GB | 941s | 72.46% | 72.70% | `w0a1g9qo` |
| B | 1,500,000 | ~80 GB | 934s | 71.37% | 71.60% | `37zg4voz` |

### Commands used

**Trial A (GPU 0):**
```bash
CUDA_VISIBLE_DEVICES=0 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --batch-size 1 --drivaerml-train-surface-points 500000 --drivaerml-eval-surface-points 500000 --num-workers 2 --max-train-batches 394 --wandb-group radford-drivaerml-round1 --wandb-name "nezuko/drivaerml-500k-pts" --agent nezuko
```

**Trial B (GPU 1):**
```bash
CUDA_VISIBLE_DEVICES=1 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --batch-size 1 --drivaerml-train-surface-points 1500000 --drivaerml-eval-surface-points 1500000 --num-workers 2 --max-train-batches 394 --wandb-group radford-drivaerml-round1 --wandb-name "nezuko/drivaerml-1500k-pts" --agent nezuko
```

### What happened

- **Both trials converged to ~71-73% relative L2**, far from the 3.71% target. The model learned almost nothing in 2 epochs.
- **1.5M pts/view performed slightly better** (71.6% vs 72.7%), suggesting more context per view helps.
- **LR barely warmed up**: With `cosine_t_max=30` and only 2 epochs, LR reached 0.000417 (vs max 0.0005). The cosine schedule hasn't even peaked.
- **2 epochs is ~1000x too few**: Published baselines use 500 epochs. Our `SENPAI_MAX_EPOCHS=2` constraint means the model has no chance to converge on DrivAerML.
- **Training throughput**: Even with `max-train-batches=394` (one random view per case), each epoch + validation took ~8 min. A full epoch without batching limits would take ~60+ min.

### Why the original trials failed

1. **First OOM**: default `drivaerml_train_surface_points=0` sends all 8.8M pts to GPU → 91+ GB needed → OOM
2. **Then infeasible iteration count**: Setting 8K-16K point caps created 400K+ iterations per epoch → stuck loading data for hours
3. **Architecture mismatch**: `geometry_supernodes` / `surface_anchor_points` don't affect `senpai_transolver`

### GPU memory benchmarks (senpai_transolver, bs=1, 3 layers, 192 hidden)

| pts/view | Peak GPU (GB) | fwd+bwd time |
|---|---|---|
| 8K | 0.5 | 0.23s |
| 50K | 2.7 | 0.03s |
| 200K | 10.7 | 0.12s |
| 500K | 26.7 | 0.30s |
| 1M | 53.3 | 0.59s |
| 1.2M | 63.9 | 0.89s |
| 1.5M | 80.1 | 0.89s |
| 2M+ | OOM | — |

### Suggested follow-ups

1. **Use `reference_abupt` model** for the anchor budget sweep — it actually uses `geometry_supernodes` and `surface_anchor_points`
2. **Increase epoch/timeout limits for DrivAerML** — 2 epochs is fundamentally insufficient for a 3D dataset this large
3. **Optimize DrivAerML data pipeline** — `num_workers=2` helped, but the NFS mmap load (~0.6s/case) is a bottleneck. Pre-caching or local SSD staging would help
4. **Consider `max_train_batches` as a permanent DrivAerML setting** — without it, a single epoch over all point views takes hours
5. **Test 1M-1.5M pts/view as default** — fits in 96GB GPUs and gives the model maximum context per view

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2439 because ADVISOR: Closing — 71.37% does not beat the 71.35% baseline (and now superseded by violet's 56.91%, #2467).

**Critical discovery retained:** geometry_supernodes and surface_anchor_points are NO-OPS for senpai_transolver — they only work with ABUPTCollate/reference_abupt. This invalidates the original anchor-budget hypothesis but reveals important architectural information. The surface point budget trend (more points = better) is valid but marginal at 2 epochs.

Thank you nezuko — the architecture discovery is genuinely valuable for future DrivAerML work..
